### ISSUE TITLE
Mermaid support

### ISSUE THREAD

**Issue by rufuspollock** [flowershow/flowershow#12]
When creating markdown pages I want to add mermaid-syntax diagrams so that I can display nicely rendered diagrams 

https://mermaid-js.github.io/mermaid/#/

## Acceptance

* [x] I can have have mermaid blocks and they render correctly **✅2022-08-16 see https://flowershow.app/docs/mermaid**

```
```mermaid
graph TD

a --> b
\```
```

## Tasks

* [x] research options e.g. existing remark plugins
* [x] Select one to try
* [x] Try it
* [x] Deploy
* [x] Document
* [x] ...

## Notes

We want a solution that renders client side (we do *not* want to render svg and insert)

Options

* https://github.com/temando/remark-mermaid **2022-07-26 this looks pretty promising for what we want**
  **2022-08-02** *summary: adding as a plugin in remark just throws error as seen in notes below*
* https://github.com/sjwall/mdx-mermaid - Plug and play Mermaid in MDX
  **2022-08-02** *summary:*
  - adding as a plugin in remark just outputs the component as a string.
  - Adding the component in MDX components client side ie. `Pre.js` is complicated as `rehype-prism-plus` converts the graph to span elements and `<Mermaid char={...} />` needs the graph as a string.
* https://github.com/remcohaszing/remark-mermaidjs - A remark plugin to render mermaid diagrams using puppeteer
   **2022-08-02** *summary: not considering as puppeteer is a large dependency even if it works without much configuration needed*

### remark-mermaid

 - throwing errors: `Expected usable value, not 'undefined'` Looks like it's trying to parse wrong values.
 
  ![Screen Shot 2022-07-26 at 6 00 07 PM](https://user-images.githubusercontent.com/42637597/181024708-2fa0d6ca-6652-494d-9a28-db358fbc8188.png)

### mdx-mermaid with remark-code-extra (UPDATE)

Using this remark plugin - https://github.com/s0/remark-code-extra - to replace `pre` tag elements created by `rehype-prism-plus`, I have managed to make `mdx-mermaid` work nicely with it. See screenshots below for result and modified code in `contentlayer.config.js`

*Brief summary of modified code:*

Transforming the nodes for `<Pre>` elements and removing the ones created by rehype-prism-plus ...

- Added the remark plugin in contentlayer with following configuration within it:
   - create a new pre tag element with just the string value as child and a new className `.code-mermaid`
   - remove the previous pre tag element created by rehype-prism-plus
- Modified the `Pre.js` component to check for the new className (code-mermaid) from props and if so, use Mermaid component ie. `<Mermaid chart={children} />`

![Screen Shot 2022-08-02 at 6 07 07 PM](https://user-images.githubusercontent.com/42637597/182394826-d5532f0a-4eea-4bdf-8c4c-e60f0093fc06.png)

The remark-code-extra plugin in `contentlayer.config.js` looks like this:

![Screen Shot 2022-08-02 at 6 18 49 PM](https://user-images.githubusercontent.com/42637597/182397270-da43a42b-fe6a-4e95-9c1d-9e8aa1e6e836.png)

### mdx-mermaid

- adding mdx-mermaid as a remark plugin just outputs a string as seen below. Probably happening since import statements in md are not working/resolved.
  
![Screen Shot 2022-07-27 at 5 18 24 AM](https://user-images.githubusercontent.com/42637597/181141397-7c94b864-3d34-471c-9e1e-3031e348fbfb.png)

using mermaid component from mdx-mermaid in `Pre.js` component works after removing `rehype-prism-plus` plugin
   ![Screen Shot 2022-07-27 at 5 24 20 AM](https://user-images.githubusercontent.com/42637597/181141680-d6d88b76-1088-4fdf-b67d-2d658924248a.png)

**Why `rehype-prism-plus` conflicts:**

rehype prism plus modifies all the code blocks including mermaid. for example below markdown

```markdown
graph LR
     Start --> Stop
```
    
would generate individual span elements for each line

```js
<span>graph LR</span>
<span>Start --> Stop</span>
```
    
Since mdx-mermaid component takes only a string `<Mermaid chart={'graph LR ....'} />`, we have to map each span element and join them as strings. This gets complex when there are elements within.

A better approach would be to somehow make `rehype-prism-plus` to discard modifying the language-mermaid code block.

### remark-mermaidjs

https://github.com/remcohaszing/remark-mermaidjs

   - works and probably should go with this ...
   - needs some style fixes
  ![Screen Shot 2022-07-26 at 6 12 21 PM](https://user-images.githubusercontent.com/42637597/181027626-dc7ec9ac-25da-4f3a-aa8b-2508d4c3eaad.png)

### Mermaidjs

https://mermaid-js.github.io/mermaid/#/

* If using this option we have to:
  - configure `Pre.js` component used in `MDXComponents` to make this work
  - Also have to do more nested stuff since **rehype-prism-plus** converts code blocks to components
     ```js
     if (/^language-mermaid/.test(children.props.className.toLocaleLowerCase())) {
       useEffect(() => mermaid.initialize({ startOnLoad: true }), [])
       return <div className="mermaid my-12">{children.props.children.map(el => el.props.children).join().replace(/,/g, "")} </div>
     }
     ```
  - Additional configuration for color code support required (not implemented)

![Screen Shot 2022-07-26 at 6 21 59 PM](https://user-images.githubusercontent.com/42637597/181029769-06dbe753-ff22-47ee-9a7b-71650df0794d.png)

**Comment by rufuspollock** [comment]
@khalilcodes starting to work on this.

**Comment by khalilcodes** [comment]
@rufuspollock updated notes with summary of some tried options.

**Comment by khalilcodes** [comment]
@rufuspollock updated the notes for https://github.com/sjwall/mdx-mermaid with further details of issues.

**Comment by rufuspollock** [comment]
@khalilcodes thanks. I just refactored notes section to be a bit clearer. Could you update the bulleted list at the top of the notes with a summary of current evaluation and recommendation (detail can be left out there and can go below in dedicated section for each option)

**Comment by khalilcodes** [comment]
@rufuspollock updated the top of notes with a summary

**Comment by rufuspollock** [comment]
@khalilcodes if this is working and relatively clean let's go with this, perhaps documenting clearly here the approach and what improvements we would like in future (with perhaps some comments in code e.g. in contentlayer segment - that can also ref here - explaining that this is for mermaid support). That way 6m or a year from now when we look at the code we can understand why we did what we did and how we would improve or replace it if a new remark plugin comes out.

My other question is whether we want a standalone remark-mermaid plugin like we had with remark-wiki-link-plus?
